docs/proposals/compute-resource-metrics-api.md

```diff
@@ -0,0 +1,179 @@
+# Kubernetes compute resource metrics API
+
+## Goals
+
+Provide resource usage metrics on pods and nodes on the API server to be used
+by the scheduler to improve job placement, utilization, etc. and by end users
+to understand the resource utilization of their jobs.
+
+## Current state
+
+Right now, the Kubelet exports container metrics via an API endpoint. This
+information is not gathered nor served by the Kubernetes API server.
+
+## Use cases
+
+The first user will be kubectl. The resource usage data can be shown to the
+user via a periodically refreshing interface similar to `top` on Unix-like
+systems. This info could let users assign resource limits more efficiently.
+
+```
+$ kubectl top kubernetes-minion-abcd
+Pod: monitoring-heapster-abcde
+CPU:
+    mean: 0.12 cores
+    95th: 1.10 cores
+    max:  1.91 cores
+Memory:
+    mean: 300MB (10%)
+    95th: 330MB (11%)
+    max:  395MB (13%)
+```
+
+A second user will be the scheduler. To assign pods to nodes efficiently, the
+scheduler needs to know the current free resources on each node.
+
+## Proposed endpoints
+
+	/api/v1/namespaces/myns/podMetrics/mypod/[derived|raw]
+	/api/v1/nodeMetrics/myNode/[derived|raw] 
```

Comment:
I can agree with that argument - so /nodemetrics/ or /podmetrics/ as a root
endpoint seems more reasonable (and avoid sub resources for the primary
interaction).

On Tue, Jul 28, 2015 at 10:34 PM, Eric Tune <notifications@github.com>
wrote:

> In docs/proposals/compute-resource-metrics-api.md
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/11951#discussion_r35723277>
> :
>
> > +    mean: 0.12 cores
> > +    95th: 1.10 cores
> > +    max:  1.91 cores
> > +Memory:
> > +    mean: 300MB (10%)
> > +    95th: 330MB (11%)
> > +    max:  395MB (13%)
> > +```
> > +
> > +A second user will be the scheduler. To assign pods to nodes efficiently, the
> > +scheduler needs to know the current free resources on each node.
> > +
> > +## Proposed endpoints
> > +
> > +	/api/v1/namespaces/myns/podMetrics/mypod/[derived|raw]
> > +	/api/v1/nodeMetrics/myNode/[derived|raw]
>
> Also, people will definitely want aggregations over both smaller and
> larger granularity than a Pod. e.g. one container of a pod, or a set of
> pods. For some useful types of metrics, like quantiles, you can't compute
> the metric on a set of pods only given the metric on a single pod.
>
> So that makes me think that metrics in general don't really fit the
> subresource model.
>
> To be clear, I'm not saying you should implement those other granularities
> right now, just that we should think about what the API looks like when we
> later want them.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/11951/files#r35723277>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift